readthedocs/donate/signals.py

```diff
@@ -1,9 +1,118 @@
+import random
+
 from django.dispatch import receiver
 from django.conf import settings
 from django.core.cache import cache
 
 from readthedocs.restapi.signals import footer_response
-from readthedocs.donate.models import SupporterPromo, VIEWS, CLICKS, OFFERS
+from readthedocs.donate.models import SupporterPromo, VIEWS, CLICKS, OFFERS, INCLUDE, EXCLUDE
+from readthedocs.donate.utils import get_ad_day
+
+
+PROMO_GEO_PATH = getattr(settings, 'PROMO_GEO_PATH', None)
+
+if PROMO_GEO_PATH:
+    import geoip2.database  # noqa
+    geo_reader = geoip2.database.Reader(PROMO_GEO_PATH)
+
+
+def show_to_geo(promo, country_code):
+    # Remove promo's that exclude this country.
+    for filter in promo.geo_filters.all():
+        if filter.filter_type == INCLUDE:
+            if country_code in filter.codes:
+                continue
+            else:
+                return False
+        if filter.filter_type == EXCLUDE:
+            if country_code in filter.codes:
+                return False
+
+    return True
+
+
+def choose_promo(promo_list):
+    """
+    This is the algorithm to pick which promo to show.
+
+    This takes into account how many remaining days this
+    promo has to be shown.
+
+    The algorithm is currently as such:
+
+    * Take the remaining number of views each promo has today
+    * Add them together, with each promo "assigned" to a range
+    * Pick a random number between 1 and that total
+    * Choose the ad whose range is in the chosen random number
+
+    In the future,
+    we should take into account the expected views for today
+    (The number of views from this day last week)
+    Then we can scale the "total ads sold" against that "expected views",
+    and that will give us more spread throughout the day.
+    """
+
+    promo_range = []
+    total_views_needed = 0
+    for promo in promo_list:
+        promo_range.append([
+            total_views_needed,
+            total_views_needed + promo.views_needed_today(),
+            promo
+        ])
+        total_views_needed += promo.views_needed_today()
+    choice = random.randint(0, total_views_needed)
+    for range_list in promo_range:
+        if range_list[0] <= choice <= range_list[1]:
+            return range_list[2]
```

Comment:
I find that to be much harder to read & reason about, and basically just reduces the for loop into a list comprehension. Less code doesn't mean easier to understand, and in experience usually means the opposite. 

I bet we could ask most Python devs, and it would take them at least 2-3 minutes to understand what `chosen = promo_list[bisect([bp for (_, bp) in promo_list], choice)]` does.